### ISSUE TITLE
Add option to INFO to return a json payload

### ISSUE THREAD

**Issue by ltagliamonte-dd** [apache/kvrocks#2302]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/kvrocks/issues) and found no similar issues.

### Motivation

try to monitor some specs from my service and would be better to be able to consume them in a format that's easier to consume rather than parsing strings like the kvrocks_exporter does.

### Solution

would be nice to expose somenthing like:
``` INFO `section` JSON/TXT```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

**Comment by VasuDevrani** [comment]
do we want INFO command to return output in json fomat when `INFO JSON` (JSON specified in command) otherwise plain text?

**Comment by PragmaTwice** [comment]
I think for compatibility, it should be `INFO [<section_name>] [FORMAT (TXT | JSON)]`.

**Comment by VasuDevrani** [comment]
im struggling with parsing the output to json,[ this line of ](https://github.com/VasuDevrani/kvrocks/blob/unstable/src/server/server.cc?plain=1#L1188) code outputs string which looks like

`"# Clients\r\nmaxclients:10000\r\nconnected_clients:1\r\nmonitor_clients:0\r\nblocked_clients:0\r\n"`

using `jsoncons::json::parse` throws ser_error (unsupported format) and I should not directly append the string as is to the json.

should i create a custom parsing function to convert it something like?
`{"clients":{"blocked_clients":"0\r","connected_clients":"1\r","maxclients":"10000\r","monitor_clients":"0\r"}}
`

**Comment by PragmaTwice** [comment]
TBH I cannot get your point. Why do you want to parse it as JSON? It's just not JSON.

**Comment by VasuDevrani** [comment]
> TBH I cannot get your point. Why do you want to parse it as JSON? It's just not JSON.

Maybe i misunderstood, please confirm. 

The issue wants the support of commands `INFO [<section_name>] [FORMAT (TXT | JSON)]`:

`INFO`: returns info in default plain text format.
`INFO <section>`: section info in plain text format.
`INFO all FORMAT JSON`: returns the entire info in JSON format.
`INFO <section> FORMAT JSON`: returns just section info in JSON.
`INFO <section> FORMAT TXT`: section info in plain text format.

**Comment by PragmaTwice** [comment]
Yeah, and you don't need to parse anything for implementing such behavior.

**Comment by VasuDevrani** [comment]
> Yeah, and you don't need to parse anything for implementing such behavior.

so is this approach not desirable
https://github.com/apache/kvrocks/compare/unstable...VasuDevrani:kvrocks:init_INFO

**Comment by PragmaTwice** [comment]
> > Yeah, and you don't need to parse anything for implementing such behavior.
> 
> so is this approach not desirable [unstable...VasuDevrani:kvrocks:init_INFO](https://github.com/apache/kvrocks/compare/unstable...VasuDevrani:kvrocks:init_INFO)

Yeah, it looks a bit ugly. We can have a more intuitive design for this feature.

**Comment by VasuDevrani** [comment]
> Yeah, it looks a bit ugly. We can have a more intuitive design for this feature.

i feel confused about the more intuitive design 
is this close to that https://github.com/apache/kvrocks/commit/292bdbe7d36f54ada89623006c1597cb74eed3ef?
otherwise its out of scope for me (at least for now, will have to wait for someone else's PR or if someone can guide me around the better approach).

**Comment by PragmaTwice** [comment]
> > Yeah, it looks a bit ugly. We can have a more intuitive design for this feature.
> 
> i feel confused about the more intuitive design is this close to that [292bdbe](https://github.com/apache/kvrocks/commit/292bdbe7d36f54ada89623006c1597cb74eed3ef)? otherwise its out of scope for me (at least for now, will have to wait for someone else's PR or if someone can guide me around the better approach).

https://github.com/apache/kvrocks/blob/292bdbe7d36f54ada89623006c1597cb74eed3ef/src/server/server.cc#L1211-L1229


I hope you can notice that there is a lot of repetition in your code between if and else, we should avoid this [copy-and-paste programming](https://en.wikipedia.org/wiki/Copy-and-paste_programming). For me, preventing such code from entering kvrocks codebase is my top priority.

In fact, we can see that the result returned by INFO is a format similar to 
```
std::map<SectionNameTy, std::map<KeyTy, ValueTy>>
(where ValueTy = something like std::variant<int, std::string>)
```
So we can return this format and transform it into JSON or TEXT at the end.

**Comment by PragmaTwice** [comment]
In details, we can have such small functions, like:
```
auto GetPersistenceInfo() -> std::map<KeyTy, ValueTy>;
auto GetServerInfo() -> std::map<KeyTy, ValueTy>;
auto GetPWhateverInfo() -> std::map<KeyTy, ValueTy>;
```

And then at the function that can generate all info (not real C++ code):
```
std::map<..> func_map = {{"persistence", GetPersistenceInfo}, ...}

auto all_info = func_map.filter(selected_info_sections).map(x -> x.call())

if (JSON) generateJsonOutput(all_info)
else (TXT) generateTxtOutput(all_info)
```

**Comment by VasuDevrani** [comment]
> ```
> auto GetPersistenceInfo() -> std::map<KeyTy, ValueTy>;
> auto GetServerInfo() -> std::map<KeyTy, ValueTy>;
> auto GetPWhateverInfo() -> std::map<KeyTy, ValueTy>;
> ```
we already does have some function like [GetServerInfo](https://github.com/apache/kvrocks/blob/acddc28de42ca41d64feea78691d877a2297932c/src/server/server.cc#L939) that uses string stream and returns string, should i overwrite them to return `std::map<KeyTy, ValueTy>`

**Comment by PragmaTwice** [comment]
> > ```
> > auto GetPersistenceInfo() -> std::map<KeyTy, ValueTy>;
> > auto GetServerInfo() -> std::map<KeyTy, ValueTy>;
> > auto GetPWhateverInfo() -> std::map<KeyTy, ValueTy>;
> > ```
> 
> we already does have some function like [GetServerInfo](https://github.com/apache/kvrocks/blob/acddc28de42ca41d64feea78691d877a2297932c/src/server/server.cc#L939) that uses string stream and returns string, should i overwrite them to return `std::map<KeyTy, ValueTy>`

Yeah you can update these methods.

**Comment by zweix123** [comment]
Is there anyone else working on solving this problem? If not, I would like to give it a try. If I encounter problems or cannot solve them, I will actively communicate with the community in a timely manner.

**Comment by nhancdt2602** [comment]
@jihuayu Could I take over this issue? It seems un-active for a long time.
